Add calculate_fee and calculate_fee_rate on wallet

reez opened the pull request:
<!-- Erase any parts of this template not applicable to your Pull Request. -->

### Description

~Add calculate_fee on wallet.~ 
Add calculate_fee and calculate_fee_rate on wallet.

Continuing to implement functions/types associated with [PR on bdk "Remove TransactionDetails from Wallet API"](https://github.com/bitcoindevkit/bdk/pull/1048).

### Notes to the reviewers

Open to hearing where to best place the new code in each file too (should `BdkCalculateFeeError` be placed somewhere different in the UDL, etc).

~Currently I have `BdkCalculateFeeError` with associated values, but I am ignoring associated values in the UDL, and will address that in a follow up PR with all of the other changes we need to possibly do to Error types and associating data with them.~ 
I have `CalculateFeeError` with associated values and expose them in the UDL as well.

### Changelog notice

<!-- Notice the release manager should include in the release tag message changelog -->
<!-- See https://keepachangelog.com/en/1.0.0/ for examples -->

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [ ] I've added docs for the new feature

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [ ] I'm linking the issue being fixed by this PR

thunderbiscuit reviewed (changes requested):
Nice little addition! Only small issue I think is that the name is reversed; you're exposing a new error type called `BdkCalculateFeeError`, but we usually expose the a type with the same name as in Rust and internally transform the Rust type into a `BdkSomethingSomething` type. Just a switcharoo of those will make this PR ready to go.

reez commented on bdk-ffi/src/bdk.udl:
```diff
@@ -82,6 +82,17 @@ enum BdkError {
 };
 
+[Error]
+interface CalculateFeeError {
```
using `interface` per [uniffi docs on errors with associated types](https://mozilla.github.io/uniffi-rs/udl/errors.html)

reez commented on bdk-ffi/src/error.rs:
```diff
@@ -0,0 +1,35 @@
+    }
+}
+
+impl std::error::Error for CalculateFeeError {}
```
Implementing the `std::error::Error` trait for `CalculateFeeError` allows error types to integrate with Rust's standard error handling ecosystem

reez commented on bdk-ffi/src/error.rs:
```diff
@@ -0,0 +1,35 @@
+    NegativeFee { fee: i64 },
+}
+
+impl fmt::Display for CalculateFeeError {
```
`fmt::Display` implementation for `CalculateFeeError`: 
- `write!` is used to format and append the error message to the `fmt::Formatter`.
- this formatted message is used wherever the display representation of the error is needed

reez commented:
Updates: 
- did switcharoo of `CalculateFeeError` and `BdkCalculateFeeError` per [Thunder suggestion](https://github.com/bitcoindevkit/bdk-ffi/pull/437#pullrequestreview-1787456514) 
- exposed `CalculateFeeError` with associated values in the UDL per [uniffi](https://mozilla.github.io/uniffi-rs/udl/errors.html) 
- added `calculate_fee_rate`
- added a simplified `FeeRate` constructor and removed the `from_sat_per_vb` constructor 
- updated tests across all 4 platforms    
 

 This is ready for review again, but any suggestions welcome.

thunderbiscuit commented on bdk-jvm/lib/src/test/kotlin/org/bitcoindevkit/LiveWalletTest.kt:
```diff
@@ -55,8 +55,14 @@ class LiveWalletTest {
-
         println("Txid is: ${tx.txid()}")
+
+        val txFee = wallet.calculateFee(tx)
+        println("Tx fee is: ${tx.txid()}")
+
+        val feeRate = wallet.calculateFeeRate(tx)
+        println("Tx fee rate is: ${tx.txid()} sat/vB")
+
         esploraClient.broadcast(tx)
```
Same comment as above. Small typo.

reez replied:
Definitely, updated-

thunderbiscuit commented on bdk-python/tests/test_live_wallet.py:
```diff
@@ -64,6 +64,11 @@ def test_broadcast_transaction(self):
+        fee = wallet.calculate_fee(tx)
+        print(f"Transaction Fee: {fee}")
+        fee_rate = wallet.calculate_fee_rate(tx)
+        print(f"Transaction Fee Rate: {fee_rate} sat/vB")
```
I don't think you're printing the value here but just the object. Me thinks it needs `fee_rate.as_sat_per_vb()`.

reez replied:
Definitely, miss by me, updated-

thunderbiscuit commented on bdk-ffi/src/bdk.udl:
```diff
@@ -82,6 +82,16 @@ enum BdkError {
 
+[Error]
+interface CalculateFeeError {
+  MissingTxOut(sequence<OutPoint> out_points);
+  NegativeFee(i64 fee);
```
This is the first time we use this pattern of defining errors as interfaces in the UDL. Can you confirm that the errors print out what we expect them to? Just making sure it all works well.

reez replied:
I just added two tests in the `error.rs` file to address this. Let me know if:
- those two tests make sense to you
- that is the best place to test these two errors

... or any other thoughts you had on it too. Thanks, great suggestion!

thunderbiscuit reviewed (changes requested):
Awesome work! I have a few comments but overall this is shaping up.

reez commented on bdk-android/lib/src/androidTest/kotlin/org/bitcoindevkit/LiveWalletTest.kt:
```diff
@@ -59,8 +59,14 @@ class LiveWalletTest {
+
+        val feeRate = wallet.calculateFeeRate(tx)
+        println("Tx fee rate is: ${tx.txid()} sat/vB")
+
```
Definitely, updated-

reez commented on bdk-ffi/src/types.rs:
```diff
@@ -7,8 +7,22 @@ use bdk::KeychainKind;
+impl FeeRate {
+    pub fn new(bdk_fee_rate: BdkFeeRate) -> Self {
+        FeeRate(bdk_fee_rate)
+    }
```
Indeed, great call, updated-

reez commented on bdk-ffi/src/types.rs:
```diff
@@ -7,8 +7,22 @@ use bdk::KeychainKind;
+    pub fn as_sat_per_vb(&self) -> f32 {
+        self.0.as_sat_per_vb()
+    }
+}
```
Totally! I wasn't sure exactly which ones I should open up other than `as_sat_per_vb()` so getting your opinion on just adding `sat_per_kwu()` is great. It is added now.

reez commented on bdk-ffi/src/wallet.rs:
```diff
@@ -98,6 +99,19 @@ impl Wallet {
+    pub fn calculate_fee(&self, tx: &Transaction) -> Result<u64, CalculateFeeError> {
+        self.get_wallet()
+            .calculate_fee(&tx.clone().into())
+            .map_err(|err| err.into())
```
Absolutely, my b, updated-

reez commented on bdk-ffi/src/wallet.rs:
```diff
@@ -473,7 +487,7 @@ impl TxBuilder {
         }
         if let Some(sat_per_vb) = self.fee_rate {
-            tx_builder.fee_rate(FeeRate::from_sat_per_vb(sat_per_vb));
+            tx_builder.fee_rate(BdkFeeRate::from_sat_per_vb(sat_per_vb));
```
Sounds great, added a [new issue](https://github.com/bitcoindevkit/bdk-ffi/issues/438) for it and assigned it to myself.

reez commented on bdk-ffi/src/wallet.rs:
```diff
@@ -98,6 +99,19 @@ impl Wallet {
+
+    pub fn calculate_fee(&self, tx: &Transaction) -> Result<u64, CalculateFeeError> {
+        self.get_wallet()
+            .calculate_fee(&tx.clone().into())
```
Sounds like a plan to me. I added your suggested code snippet:

```
impl From<&Transaction> for BdkTransaction {
    fn from(tx: &Transaction) -> Self { tx.inner.clone() }
}
```

... then updated 4 functions that had a line related to transaction that could have `.clone()` removed.

wallet.rs
- `sent_and_received`
- `calculate_fee`
- `calculate_fee_rate`

esplora.rs
- `broadcast`

... there was still 1 transaction related `.clone()` and it was wallet.rs `transactions` so I added the reference version (we already have the owned version):

```
impl From<&BdkTransaction> for Transaction {
    fn from(tx: &BdkTransaction) -> Self {
        Transaction { inner: tx.clone() }
    }
}
```

... and then updated the `transactions` function code to remove `.clone()`.

I tested `bdk-ffi` `bdk-swift` `bdk-jvm` locally with these changes and tests still passed.

Hopefully that all makes sense to have updated and added along with your suggestion!

reez commented:
> Awesome work! I have a few comments but overall this is shaping up.

Great suggestions/thoughts/catches. Appreciated!

I updated all of the small items (and just marked the conversation resolved since they had been updated and were now “Outdated”) hopefully that’s cool, but I left the 2 items that I updated code for but you might still want to converse about too: 
- [trait on Transaction](https://github.com/bitcoindevkit/bdk-ffi/pull/437#discussion_r1437850039). I added comments to this as well as updated code, but just leaving it open for any other feedback or thoughts you might have
- [error testing](https://github.com/bitcoindevkit/bdk-ffi/pull/437#discussion_r1437835474). I added comments and code/tests for this but just leaving it open too for any other feedback or thoughts.

reez commented:
Rebased

thunderbiscuit reviewed (approved):
ACK 3789c1dcd69827247d7a6afaf1bf36506fa311aa. Thanks!
